File: python/pn_example.py
from __future__ import annotations

from dataclasses import dataclass

from rxnet.pn import Arc, Context, Net, Runtime, Transition

P_IDLE = 0
P_BUSY = 1


@dataclass
class Inputs:
    add: int = 0
    remove: int = 0


@dataclass
class AppData:
    fired_count: int = 0


def want_add(ctx: Context, user: AppData) -> bool:
    return ctx.latched_inputs.add != 0


def want_remove(ctx: Context, user: AppData) -> bool:
    return ctx.latched_inputs.remove != 0


def on_fire(ctx: Context, user: AppData) -> None:
    user.fired_count += 1


def main() -> None:
    app = AppData()
    runtime = Runtime(inputs=Inputs())

    net = Net(
        name="queue",
        places=[1, 0],
        transitions=[
            Transition(
                consume=[Arc(P_IDLE, 1)],
                produce=[Arc(P_BUSY, 1)],
                guard=want_add,
                action=on_fire,
            ),
            Transition(
                consume=[Arc(P_BUSY, 1)],
                produce=[Arc(P_IDLE, 1)],
                guard=want_remove,
                action=on_fire,
            ),
        ],
        user=app,
    )

    runtime.add_net(net)

    print(f"init: places={net.places} fired={app.fired_count}")

    runtime.context.inputs.add = 1
    runtime.context.inputs.remove = 0
    runtime.tick()
    print(f"after add: places={net.places} fired={app.fired_count}")

    runtime.context.inputs.add = 0
    runtime.context.inputs.remove = 1
    runtime.tick()
    print(f"after remove: places={net.places} fired={app.fired_count}")


if __name__ == "__main__":
    main()

File: c/examples/pn_example.c
#include "rxnet/pn.h"

#include <stdio.h>

typedef struct {
    int add;
    int remove;
} app_inputs;

enum {
    P_IDLE = 0,
    P_BUSY = 1,
};

typedef struct {
    int fired_count;
    app_inputs *inputs;
} app_data;

static int want_add(const rx_pn_context *ctx, void *user) {
    const app_inputs *in = ((const app_data *)user)->inputs;
    return in->add != 0;
}

static int want_remove(const rx_pn_context *ctx, void *user) {
    const app_inputs *in = ((const app_data *)user)->inputs;
    return in->remove != 0;
}

static void on_fire(rx_pn_context *ctx, void *user) {
    ((app_data *)user)->fired_count += 1;
}

int main(void) {
    app_inputs inputs = {0};
    app_data app = {0, &inputs};
    rx_pn_runtime *runtime;
    rx_pn_net *net;

    int initial_places[] = {1, 0};

    rx_pn_arc t0_consume[] = {{P_IDLE, 1}};
    rx_pn_arc t0_produce[] = {{P_BUSY, 1}};

    rx_pn_arc t1_consume[] = {{P_BUSY, 1}};
    rx_pn_arc t1_produce[] = {{P_IDLE, 1}};

    rx_pn_transition transitions[] = {
        {t0_consume, 1, t0_produce, 1, want_add, on_fire},
        {t1_consume, 1, t1_produce, 1, want_remove, on_fire},
    };

    runtime = rx_pn_runtime_create(1);
    if (runtime == NULL) {
        fprintf(stderr, "rx_pn_runtime_create failed\n");
        return 1;
    }

    net = rx_pn_net_create("queue", initial_places, 2, transitions, 2, &app);
    if (net == NULL) {
        fprintf(stderr, "rx_pn_net_create failed\n");
        rx_pn_runtime_destroy(runtime);
        return 1;
    }

    if (rx_pn_runtime_add_net(runtime, net) != 0) {
        fprintf(stderr, "rx_pn_runtime_add_net failed\n");
        rx_pn_net_destroy(net);
        rx_pn_runtime_destroy(runtime);
        return 1;
    }

    printf("init: places=[%d,%d] fired=%d\n", net->places[P_IDLE], net->places[P_BUSY], app.fired_count);

    inputs.add = 1;
    inputs.remove = 0;
    rx_pn_tick(runtime);
    printf("after add: places=[%d,%d] fired=%d\n", net->places[P_IDLE], net->places[P_BUSY], app.fired_count);

    inputs.add = 0;
    inputs.remove = 1;
    rx_pn_tick(runtime);
    printf("after remove: places=[%d,%d] fired=%d\n", net->places[P_IDLE], net->places[P_BUSY], app.fired_count);

    rx_pn_net_destroy(net);
    rx_pn_runtime_destroy(runtime);
    return 0;
}
